content/source/docs/plugin/framework/validation.html.md

```diff
@@ -0,0 +1,257 @@
+
+## Syntax and Basic Schema Validation
+
+The [Terraform configuration language](https://www.terraform.io/docs/language/) is declarative and an implementation of [HashiCorp Configuration Language](https://github.com/hashicorp/hcl) (HCL). Terraform CLI is responsible for reading and parsing configurations for validity, based on Terraform's concepts such as `resource` blocks and associated syntax. Basic validation of value type and behavior information, for example returning an error when a string value is given where a list value is expected or returning an error when a required attribute is missing from a configuration, is automatically handled by Terraform CLI based on the provider, resource, or data source schema.
+
+Any further validation provided by the framework occurs after these checks.
+
+## Attribute Validation
+
+It is common for provider implementations to introduce validation on attributes using the generic framework-defined types such as `types.String`. The [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) can be supplied with a list of validations and diagnostics will be returned from all validators. For example:
+
+```go
+// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
+// resource, or data source.
+tfsdk.Attribute{
+    // ... other Attribute configuration ...
+
+    Validators: []AttributeValidators{
+        // These are example validators
+        stringLengthBetween(10, 256),
+        stringRegularExpression(regexp.MustCompile(`^[a-z0-9]+$`)),
+    },
+}
+```
+
+-> The framework will implement or reference common use case attribute validations, such as string length, in the future.
+
+### Creating Attribute Validators
+
+To create an attribute validator, the [`tfsdk.AttributeValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator) must be satisfied. For example:
+
+```go
+type stringLengthBetweenValidator struct {
+    Max int
+    Min int
+}
+
+func (v stringLengthBetweenValidator) Description(ctx context.Context) string {
+    return fmt.Sprintf("string length must be between %d and %d", v.Min, v.Max)
+}
+
+func (v stringLengthBetweenValidator) MarkdownDescription(ctx context.Context) string {
+    return fmt.Sprintf("string length must be between `%d` and `%d`", v.Min, v.Max)
+}
+
+func (v stringLengthBetweenValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+    str, ok := request.AttributeConfig.(types.String)
+
+    if !ok {
+        response.Diagnostics.AddAttributeError(
+            request.AttributePath,
+            "Invalid Attribute Value Type",
+            fmt.Sprintf("Expected types.String value, got: %T", request.AttributeConfig),
+        )
+
+        return
+    }
```

Comment:
```suggestion
    // types.String must be the attr.Value produced by the attr.Type in the schema for this attribute
    // for generic validators, use
    // https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ConvertValue
    // to convert into a known type.
    var str types.String
    diags := tfsdk.ValueAs(ctx, val, &target)
    resp.Diagnostics.Append(diags...)
    if resp.Diagnostics.HasError() {
        return
    }
```

This explains that we can only use the `attr.Value` produced by the `attr.Type` in the schema, tells people how to convert between `attr.Type`s, and uses the `tfsdk.ValueAs` instead of type assertion to get the value.

Reply:
Love it, just needs minor variable name fixes.

```suggestion
    // types.String must be the attr.Value produced by the attr.Type in the schema for this attribute
    // for generic validators, use
    // https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#ConvertValue
    // to convert into a known type.
    var str types.String
    diags := tfsdk.ValueAs(ctx, req.AttributeConfig, &str)
    resp.Diagnostics.Append(diags...)
    if diags.HasError() {
        return
    }
```